File: CS480/senior_project/VideoRecipeGeneratorServer/src/geminiParser.py
from google.cloud import aiplatform
import sys


# https://cloud.google.com/vertex-ai/docs/python-sdk/use-vertex-ai-python-sdk

# https://console.cloud.google.com/vertex-ai/publishers/google/model-garden/gemini-pro
import vertexai
from vertexai.preview.generative_models import GenerativeModel, ChatSession

# project_id = "PROJECT_ID"
location = "us-central1"
vertexai.init(project="gen-lang-client-0030813916")

model = GenerativeModel("gemini-1.5-flash-002")
chat = model.start_chat()


def get_chat_response(chat: ChatSession, prompt: str):
    response = chat.send_message(prompt)
    return response.text


transcript = sys.argv[1]
fileID = sys.argv[2]
prompt = (
    "From the following transcript of a recipe video, generate a recipe with an ingredients section and an instructions section: "
    + transcript
    + ". Please use the following JSON schema: "
    + '{"name": "string", "ingredients": [{"ingredient": "string", "amount": "string"}], "instructions": ["string"]}. If any information is missing, substitute with "BLANK". Thank you!'
)

stream = open("./outputFiles/parseFiles/parse-" + fileID + ".json", "w")

toWrite = get_chat_response(
    chat,
    prompt,
)


def scrub_lines(text):
    lines = text.strip().split("\n")
    answer = []

    for index, line in enumerate(lines):
        if index == 0 or index == len(lines) - 1:
            continue
        else:
            answer.append(line)
    return answer


toWrite = scrub_lines(toWrite)
toWrite = "\n".join(toWrite)
stream.write(toWrite)

File: CS480/senior_project/VideoRecipeGeneratorServer/src/geminiParser.js
import { GoogleGenerativeAI } from "@google/generative-ai";
import fs from "fs";

class GeminiParser {
  #genAI;
  #model;
  constructor() {
    try {
      this.#genAI = new GoogleGenerativeAI(process.env.GOOGLE_KEY); //
      this.#model = this.#genAI.getGenerativeModel({
        model: "gemini-1.5-flash",
      });
    } catch (error) {
      console.error(error);
    }
  }
  async generateParse(aTranscript, outputFileId) {
    return new Promise(async (resolve, reject) => {
      try {
        if (
          fs.existsSync(`./outputFiles/parseFiles/parse-${outputFileId}.txt`)
        ) {
          console.log(
            `geminiParser.js: ./outputFiles/parseFiles/parse-${outputFileId}.txt exists already, resolving with that`
          );
          return resolve(`./outputFiles/parseFiles/parse-${outputFileId}.txt`);
        }

        let fileId = outputFileId;
        let prompt = this.generatePrompt(aTranscript);
        let result = await this.#model.generateContent(prompt);
        fs.writeFileSync(
          `./outputFiles/parseFiles/parse-${fileId}.txt`,
          this.scrubLines(result.response.text())
        );

        let outputPath = `./outputFiles/parseFiles/parse-${fileId}.txt`;
        return resolve(outputPath);
      } catch (e) {
        console.error(`geminiParser.js: ${e}`);
        return reject(e);
      }
    });
  }
  generatePrompt(aTranscript) {
    return (
      "From the following transcript of a recipe video, generate a recipe with an ingredients section and an instructions section: ```" +
      aTranscript +
      "``` Please use the following JSON schema: " +
      JSON.stringify({
        name: "string",
        ingredients: [{ ingredient: "string", amount: "string" }],
        instructions: [{ instruction: "string" }],
      }) +
      `. If any information is missing, replace it with the string "BLANK". Ensure the instructions array contains objects, where each object has an instruction string without numbers, bullet points, or additional formatting.
      
            Here's an example output:
            {
              "name": "Pasta Recipe",
              "ingredients": [
                { "ingredient": "Pasta", "amount": "200 grams" },
                { "ingredient": "Tomato Sauce", "amount": "1 cup" }
              ],
              "instructions": [
                { "instruction": "Boil the pasta until cooked." },
                { "instruction": "Mix the pasta with the tomato sauce." }
              ]
            }
            
            Thank you!`
    );
  }
  scrubLines(text) {
    //necessary bc Gemini for some reason includes markdown backticks for generating a code block in its output
    let ans = "";
    let lines = text.split("\n");
    for (let i = 0; i < lines.length; i++) {
      if (i != 0 && i != lines.length - 2) {
        ans += lines[i] + "\n";
      }
    }
    return ans;
  }
}
export default GeminiParser;
